### ISSUE TITLE
input for RNN (LSTM) when using MEL-coefficients for speech recognition

### ISSUE THREAD

**Issue by InterTriplete2010** [mlpack/models#42]
Ok, so I have another question related to the input of the neural network. This time is about RNN (LSTM). I am trying to do some simple speech recognition using MEL-coefficients. I am trying to adapt the example of the stock prediction (https://github.com/mlpack/examples/blob/master/lstm_stock_prediction/lstm_stock_prediction.cpp) to my needs, but I am having problems figuring out how to properly set-up my 3D matrix. 

Say that I have a matrix of size 2000x1000, where each row represents a different audio file (let's assume we have 10 words to recognize => 200 audio files per word) and each column represents a sample of a MEL-coefficient. Let's assume that I have 20 MEL-coefficients, each one comprising of 50 samples. So basically the first 50 samples represent the samples of the first MEL coefficient, the next 50 samples represent the samples of the second MEL coefficients, etc. The MEL coefficients are also organized in ascending order.   

How do my 3D Train and Test matrix should look like, assuming (for instance), that I am using 10% of the words for validation? 

Thank you for your help!!!
Alex.

**Comment by rcurtin** [comment]
Hi @InterTriplete2010, sorry for the slow response on this one!  I believe that the shape that you are looking for will be an `arma::cube` of 20 rows x 2000 columns x 50 slices.  That is, each row represents a dimension of a single sample.  So in your case you say that there are 20 Mel coefficients in each sample.  Thus, you would have 20 rows.  Each column should represent a sequence/audio file (just like how each column in all mlpack methods represents an observation).  Then, each time step should be a slice---and based on what you wrote I think there are 50 time steps in your data.

Anyway, hopefully this clarifies the general idea.  You might have to do some restructuring of your data to get it into that format. :+1:

**Comment by InterTriplete2010** [comment]
Thank you 👍

**Comment by InterTriplete2010** [comment]
Hi @rcurtin,

Sorry to ask this question in a thread that has already been closed, but I got caught up with other things and only last week I was able to get back to work with RNN. 

The input layer is now very clear (thank you for the explanation). One other thing that I am having problems to wrap my head around is how to set-up the target labels with the MEL coefficients. I thought the arma::cube for the labels should be something like this: "1 x N sample x 1", since I am trying to get a single output for each data point. However, the code crashes unless my third dimension is identical to the number of slices, so I should have something like 1 x N x 50 with an input layer of 20 x 2000 x 50, like in my previous examples. 

Now, I understand the need to have this format in one of the examples reported in github (e.g. stock price), but in my case I am trying to predict single words, so why would I need all of the 50 time slices for the target labels if my goal is to have a single output for each of the 13 features x slices of each sample? I am still new to speech recognition with DNN, so I am sure I am missing something here, but I cannot figure out what. I spent hours browsing internet before deciding to bother you one more time. If you could help me figure out what is wrong in my way of thinking, I would be very grateful. Thank you so much!!! Alex.

**Comment by InterTriplete2010** [comment]
Sorry, just adding one more thing in order to see if I am approaching the problem in the correct way. I am trying to solve a Many-to-One Sequence problem here, so based on what I read on Keras (I know each package works in different way, but I figured it doesn't hurt to see how they operate) the arma::cube for the labels should be how I described it above, that is a single vector with N columns.

**Comment by InterTriplete2010** [comment]
Sorry, just adding that I found a similar example in keras (I know each package is doing things differently, but I figures it won't hurt to check it) and they have as vector as input for the labels and not a 3D matrix.
